rfd/0179-api-versioning.md

```diff
@@ -0,0 +1,392 @@
+
+authors: Michael Myers (michael.myers@goteleport.com)
+state: draft
+---
+
+# RFD 179 - Web API backward compatibility and typing system
+
+## What
+
+Discussion around viable options to help version and type our evolving web API
+
+## Why
+
+The Web API/Client have generally been seen as one-in-the-same with the proxy
+service. While that is somewhat true due to the fact that the proxy serves the
+web client, there are scenarios where multiple versions of the proxy can exist
+in a cluster. This means that the web api should be thought of as a separate
+"component" when discussing our [backward compatibility promise](https://goteleport.com/docs/upgrading/overview/#component-compatibility).
+Also, the types for our api are completely opaque between the client code and
+the server code. This leads to guessing json responses, unnecessary type
+conversions on the client and server, and lack of guardrails around the changing
+of request/response types. It is not only bad ergonomics for developers, but
+leaves the door wide open for breaking backward compatibility.  
+
+## Example
+We recently updated a few resources to be paginated in the web UI instead of
+sending the entire list. This example is taken from a recent feature update to
+the Access Requests list in the Web UI and highlights the issue at hand.
+
+Access Requests in the web UI used a client side paginated table, and would fetch every request in the initial API call.
+```go
+// example handler
+func AccessRequestHandler() ([]AccessRequest, error) {
+    accessRequests := []AccessRequest
+    return accessRequests,nil
+}
+```
+ Some users had too many access requests to send over the fetch and the page
+ would break, becoming unusable for larger users. The solution was to server
+ side paginate the Access Requests request. The return used to just be a list of
+ Access Requests, and it has then changed to an object with a list of Access
+ Requests as a field and a nextKey field.
+ 
+ ```go
+// example handler
+func AccessRequestHandler() ([]AccessRequest, error) {
+    accessRequests := []AccessRequest
+    nextKey := "hi_im_next"
+    return AccessRequestResponse{
+        requests: accessRequests,
+        nextKey: nextKey,
+    }, nil
+}
+```
+
+This breaks backward compatibility because a web client of version n-1 would be
+expecting a list of requests only, and a web client of n hitting a proxy of n-1
+would be expecting to receive the paginated list.
+
+The solution _in this particular case_ was
+1. on the proxy side, check which version of the web client was making the request (by the existence of a specific query param, `limit`) and decide which shape to send
+2. on the web client side, if receiving a response from an old proxy, fallback to the client pagination.
```

Comment:
Yeah this one is a weird one, but definitely a valid concern. A few customers of ours specifically have run into issues running multiple versions of the proxy in a rolling upgrade, which causes major issues for them when the web client (served by proxy N) hits a proxy N-1 in their load balancer. 

It's one of the reasons for the JavaScript bundle condense project earlier this year because web clients couldn't even download the right bundle files properly when split. This happened with API calls as well

It's essentially an edge case, but large enough to consider.

Reply:
It might be worth it to explicitly state it out in the RFD, since it's not a compatibility guarantee that's typically thought about in the context of any other client <-> server relationship within Teleport.

Reply:
>A few customers of ours specifically have run into issues running multiple versions of the proxy in a rolling upgrade, which causes major issues for them when the web client (served by proxy N) hits a proxy N-1 in their load balancer.

I'm thinking how versioning will help with this. Let's assume the example from above. The developer adds a new `v2/webapi/accessrequests` endpoint that returns a paginated list of resources. This endpoint is called by the new version of the frontend. Now, when a rolling update is in progress, we may have two versions of the proxy. What should happen if we hit the older proxy with the new frontend and get 404? 
- Show the error to the user? But this means that we will end up with a broken service (I know that you mentioned handling 404 in the other paragraph, but that one was about a new feature, no?)
- Retry the call but to `v1/*` and then do that manual conversion to the paginated list on the JS side? 

Probably the ideal solution for such cases would be to have a load balancer that routes the requests based on the required version. Unfortunately, I don't think it's something we can require from customers (+ configuring this would be non-trivial) :(

Reply:
> Show the error to the user? But this means that we will end up with a broken service (I know that you mentioned handling 404 in the other paragraph, but that one was about a new feature, no?)

New feature yes but in this scenario it also works. Think of leaf clusters as well, they may be behind a version or two and could have the same issue (such as pinned resources). Most users don't do this rolling upgrade, so they won't see it. and I think a feature (either updated or new) showing a message saying "hey this doesn't work correctly until its all upgraded" isn't a bad thing. It lets the users know "teleport isn't randomly broken, its broken because of _this_" is better than random 404 and also better than trying to call an older version of a web api which, after these guidelines, is only there because it could _not_ support this new update.

Reply:
> Unfortunately, I don't think it's something we can require from customers

yeah agreed. it was an idea we had when we were trying to solve the app.js bundle issue and realized it was too out of our hands